File: packages/sdk-python/src/openllm/_call.py
"""``llm_call`` context manager: the per-request handle every Python integration
uses to record a single LLM operation.

Parity contract — identical surface to the .NET, Node.js, and Go SDKs:

* ``with openllm.llm_call(provider, model, route, tenant, team, app, env, project) as op:``
* ``op.set_prompt_tokens(int)`` / ``op.set_completion_tokens(int)``
* ``op.set_error_kind(str | None)``
* ``op.set_usage_dollars(float | None)``

On exit the context manager:

1. Emits the OTel histogram ``gen_ai.client.operation.duration``.
2. Emits the OTel counter ``gen_ai.client.token.usage`` once per token type
   (``input`` / ``output``).
3. Emits the ``llm_requests_total`` counter labelled by provider, model,
   route, tenant, team, app, env, project, and error_kind.
4. Emits ``llm_usage_dollars`` if the caller supplied a USD amount.
5. Closes the span ``chat <model>`` (or ``<operation> <model>``).
6. Detaches tenant baggage attached on entry.

The handle never accepts prompt or completion text; only counts and metadata.
"""

from __future__ import annotations

import time
from types import TracebackType
from typing import Any

from opentelemetry.trace import Span, SpanKind, Status, StatusCode

from ._baggage import attach_tenant_baggage, detach_tenant_baggage
from ._init import _get_runtime
from ._semconv import (
    DEFAULT_OPERATION,
    GEN_AI_ERROR_TYPE,
    GEN_AI_OPERATION_NAME,
    GEN_AI_REQUEST_MODEL,
    GEN_AI_SYSTEM,
    GEN_AI_TOKEN_TYPE,
    LLM_APP,
    LLM_ENV,
    LLM_ERROR_KIND,
    LLM_MODEL,
    LLM_PROJECT,
    LLM_PROVIDER,
    LLM_ROUTE,
    LLM_TEAM,
    LLM_TENANT,
    TOKEN_TYPE_INPUT,
    TOKEN_TYPE_OUTPUT,
)


class LLMCall:
    """Active LLM operation. Returned by :func:`llm_call` as the ``with`` target.

    Mutate the handle while inside the ``with`` block; the SDK reads the final
    state on ``__exit__`` and emits metrics + closes the span. Setters are
    idempotent — the last value supplied wins.
    """

    __slots__ = (
        "_provider",
        "_model",
        "_route",
        "_tenant",
        "_team",
        "_app",
        "_env",
        "_project",
        "_operation",
        "_prompt_tokens",
        "_completion_tokens",
        "_error_kind",
        "_usage_dollars",
        "_span",
        "_baggage_token",
        "_start_ns",
        "_closed",
    )

    def __init__(
        self,
        *,
        provider: str,
        model: str,
        route: str = "",
        tenant: str = "",
        team: str = "",
        app: str = "",
        env: str = "",
        project: str = "",
        operation: str = DEFAULT_OPERATION,
    ) -> None:
        self._provider = provider
        self._model = model
        self._route = route
        self._tenant = tenant
        self._team = team
        self._app = app
        self._env = env
        self._project = project
        self._operation = operation or DEFAULT_OPERATION
        self._prompt_tokens: int | None = None
        self._completion_tokens: int | None = None
        self._error_kind: str | None = None
        self._usage_dollars: float | None = None
        self._span: Span | None = None
        self._baggage_token: object | None = None
        self._start_ns: int = 0
        self._closed = False

    # ------------------------------------------------------------------
    # Setters mutated by the caller inside the ``with`` block.
    # ------------------------------------------------------------------

    def set_prompt_tokens(self, count: int) -> None:
        """Record the number of input/prompt tokens for this call."""
        if count is None:
            return
        self._prompt_tokens = max(int(count), 0)

    def set_completion_tokens(self, count: int) -> None:
        """Record the number of output/completion tokens for this call."""
        if count is None:
            return
        self._completion_tokens = max(int(count), 0)

    def set_error_kind(self, kind: str | None) -> None:
        """Set the normalized error category, or ``None`` to clear it."""
        if kind is None or kind == "":
            self._error_kind = None
        else:
            self._error_kind = str(kind)

    def set_usage_dollars(self, dollars: float | None) -> None:
        """Record an estimated USD cost for the call, or ``None`` to skip."""
        if dollars is None:
            self._usage_dollars = None
        else:
            self._usage_dollars = max(float(dollars), 0.0)

    # ------------------------------------------------------------------
    # Context manager protocol.
    # ------------------------------------------------------------------

    def __enter__(self) -> "LLMCall":
        runtime = _get_runtime()
        self._baggage_token = attach_tenant_baggage(
            tenant=self._tenant or None,
            team=self._team or None,
            app=self._app or None,
            env=self._env or None,
            project=self._project or None,
        )

        span_name = f"{self._operation} {self._model}".strip()
        span_attributes: dict[str, Any] = {
            GEN_AI_SYSTEM: self._provider,
            GEN_AI_REQUEST_MODEL: self._model,
            GEN_AI_OPERATION_NAME: self._operation,
        }
        if self._tenant:
            span_attributes[LLM_TENANT] = self._tenant
        if self._team:
            span_attributes[LLM_TEAM] = self._team
        if self._app:
            span_attributes[LLM_APP] = self._app
        if self._env:
            span_attributes[LLM_ENV] = self._env
        if self._project:
            span_attributes[LLM_PROJECT] = self._project
        if self._route:
            span_attributes[LLM_ROUTE] = self._route

        self._span = runtime.tracer.start_span(
            name=span_name,
            kind=SpanKind.CLIENT,
            attributes=span_attributes,
        )
        self._start_ns = time.perf_counter_ns()
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None:
        if self._closed:
            return
        self._closed = True

        elapsed_ns = time.perf_counter_ns() - self._start_ns
        duration_seconds = elapsed_ns / 1_000_000_000

        # If the caller did not set an error kind but an exception is in
        # flight, derive a default kind from the exception class name.
        if exc is not None and self._error_kind is None:
            self._error_kind = type(exc).__name__

        try:
            self._record_metrics(duration_seconds)
        finally:
            span = self._span
            if span is not None:
                if self._prompt_tokens is not None:
                    span.set_attribute(
                        "gen_ai.usage.input_tokens", int(self._prompt_tokens)
                    )
                if self._completion_tokens is not None:
                    span.set_attribute(
                        "gen_ai.usage.output_tokens", int(self._completion_tokens)
                    )
                if self._error_kind is not None:
                    span.set_attribute(GEN_AI_ERROR_TYPE, self._error_kind)
                    span.set_status(Status(StatusCode.ERROR, self._error_kind))
                    if exc is not None:
                        span.record_exception(exc)
                else:
                    span.set_status(Status(StatusCode.OK))
                span.end()

            token = self._baggage_token
            self._baggage_token = None
            if token is not None:
                detach_tenant_baggage(token)

    # ------------------------------------------------------------------
    # Internal: metric emission.
    # ------------------------------------------------------------------

    def _genai_attributes(self) -> dict[str, str]:
        attrs: dict[str, str] = {
            GEN_AI_SYSTEM: self._provider,
            GEN_AI_REQUEST_MODEL: self._model,
            GEN_AI_OPERATION_NAME: self._operation,
        }
        if self._error_kind is not None:
            attrs[GEN_AI_ERROR_TYPE] = self._error_kind
        return attrs

    def _llm_attributes(self) -> dict[str, str]:
        return {
            LLM_PROVIDER: self._provider,
            LLM_MODEL: self._model,
            LLM_ROUTE: self._route,
            LLM_TENANT: self._tenant,
            LLM_TEAM: self._team,
            LLM_APP: self._app,
            LLM_ENV: self._env,
            LLM_PROJECT: self._project,
            LLM_ERROR_KIND: self._error_kind or "",
        }

    def _record_metrics(self, duration_seconds: float) -> None:
        runtime = _get_runtime()

        genai_attrs = self._genai_attributes()
        runtime.operation_duration.record(duration_seconds, attributes=genai_attrs)

        if self._prompt_tokens:
            runtime.token_usage.add(
                int(self._prompt_tokens),
                attributes={**genai_attrs, GEN_AI_TOKEN_TYPE: TOKEN_TYPE_INPUT},
            )
        if self._completion_tokens:
            runtime.token_usage.add(
                int(self._completion_tokens),
                attributes={**genai_attrs, GEN_AI_TOKEN_TYPE: TOKEN_TYPE_OUTPUT},
            )

        llm_attrs = self._llm_attributes()
        runtime.requests_total.add(1, attributes=llm_attrs)
        if self._usage_dollars is not None and self._usage_dollars > 0:
            runtime.usage_dollars.add(
                float(self._usage_dollars), attributes=llm_attrs
            )


def llm_call(
    provider: str,
    model: str,
    route: str = "",
    tenant: str = "",
    team: str = "",
    app: str = "",
    env: str = "",
    project: str = "",
    operation: str = DEFAULT_OPERATION,
) -> LLMCall:
    """Open a new :class:`LLMCall` handle to instrument one LLM operation.

    Use as a context manager::

        with openllm.llm_call(
            provider="openai",
            model="gpt-4o-mini",
            route="primary",
            tenant="acme",
            team="growth",
            app="chatbot",
            env="prod",
            project="customer-support",
        ) as op:
            response = openai_client.chat.completions.create(...)
            op.set_prompt_tokens(response.usage.prompt_tokens)
            op.set_completion_tokens(response.usage.completion_tokens)

    The handle never accepts prompt or completion text.
    """
    if not provider:
        raise ValueError("openllm.llm_call: provider must be a non-empty string")
    if not model:
        raise ValueError("openllm.llm_call: model must be a non-empty string")
    return LLMCall(
        provider=provider,
        model=model,
        route=route,
        tenant=tenant,
        team=team,
        app=app,
        env=env,
        project=project,
        operation=operation,
    )


__all__ = ["LLMCall", "llm_call"]

File: packages/sdk-go/call.go
// Copyright 2026 Yasvanth Udayakumar
// Licensed under the Apache License, Version 2.0.

package openllm

import (
	"context"
	"sync"
	"time"

	"go.opentelemetry.io/otel/attribute"
	"go.opentelemetry.io/otel/codes"
	"go.opentelemetry.io/otel/metric"
	"go.opentelemetry.io/otel/trace"
)

// LlmCall is the handle returned by StartLlmCall. It accumulates token counts
// and dollarized usage during the call, then emits the histogram, counters,
// and span when End is invoked.
//
// LlmCall is not safe for concurrent use by multiple goroutines — instrument a
// single LLM call from a single goroutine. If you have to fan out, start a
// child call per goroutine.
//
// End is idempotent: calling it twice is a no-op on the second invocation.
type LlmCall struct {
	rt        *runtime
	span      trace.Span
	opts      CallOptions
	startedAt time.Time

	promptTokens     int64
	completionTokens int64
	hasUsageDollars  bool
	usageDollars     float64
	errorKind        string

	once sync.Once
}

// StartLlmCall opens a span for an LLM call and returns a handle plus a
// derived context that carries the span and tenant baggage. The caller MUST
// pass the returned context to downstream HTTP/SDK calls so trace context and
// tenant baggage propagate. Always defer op.End().
//
// If Init has not been called (or has been shut down) StartLlmCall returns a
// no-op handle and the input context unchanged — instrumentation must never
// break the caller.
func StartLlmCall(ctx context.Context, opts CallOptions, fnOpts ...CallOption) (*LlmCall, context.Context) {
	for _, fn := range fnOpts {
		fn(&opts)
	}
	if opts.Operation == "" {
		opts.Operation = DefaultOperation
	}

	rt, err := getRuntime()
	if err != nil {
		// No-op handle. End() will check rt == nil and skip emission.
		return &LlmCall{opts: opts, startedAt: time.Now()}, ctx
	}

	// Apply tenant baggage BEFORE starting the span so the span snapshot can
	// reflect the baggage set on the context.
	ctx = attachTenantBaggage(ctx, opts)

	spanAttrs := buildSpanAttributes(opts, rt.defaultTags)
	ctx, span := rt.tracer.Start(ctx, SpanNameLlmCall,
		trace.WithSpanKind(trace.SpanKindClient),
		trace.WithAttributes(spanAttrs...),
	)

	return &LlmCall{
		rt:        rt,
		span:      span,
		opts:      opts,
		startedAt: time.Now(),
	}, ctx
}

// SetPromptTokens records the number of prompt (input) tokens consumed by the
// call. Negative values are ignored.
func (c *LlmCall) SetPromptTokens(n int64) {
	if c == nil || n < 0 {
		return
	}
	c.promptTokens = n
}

// SetCompletionTokens records the number of completion (output) tokens.
func (c *LlmCall) SetCompletionTokens(n int64) {
	if c == nil || n < 0 {
		return
	}
	c.completionTokens = n
}

// SetErrorKind tags the call with a normalized error category (e.g.
// "rate_limit", "auth", "timeout"). Empty preserves success-state.
func (c *LlmCall) SetErrorKind(kind string) {
	if c == nil {
		return
	}
	c.errorKind = kind
}

// SetUsageDollars records dollarized usage for the call, surfaced on the
// llm_usage_dollars counter. Provided by the caller because runtime-side cost
// estimation is owned by F017, not by the SDK.
func (c *LlmCall) SetUsageDollars(amount float64) {
	if c == nil {
		return
	}
	c.usageDollars = amount
	c.hasUsageDollars = true
}

// End closes the span and emits the operation-duration histogram, the
// token-usage counters (split by gen_ai.token.type), the llm_requests_total
// counter, and the llm_usage_dollars counter (when set).
//
// End is idempotent; the second call is a no-op.
func (c *LlmCall) End() {
	if c == nil {
		return
	}
	c.once.Do(func() {
		if c.rt == nil || c.span == nil {
			// No-op handle (Init not called, or shut down before End).
			return
		}

		duration := time.Since(c.startedAt)
		ctx := trace.ContextWithSpan(context.Background(), c.span)

		genAIAttrs := genAIAttributeSet(c.opts, c.errorKind)
		c.rt.clientOpDuration.Record(ctx, duration.Seconds(),
			metric.WithAttributes(genAIAttrs...),
		)

		if c.promptTokens > 0 {
			attrs := append(append([]attribute.KeyValue{}, genAIAttrs...),
				attribute.String(GenAITokenType, TokenTypeInput))
			c.rt.clientTokenUsage.Add(ctx, c.promptTokens, metric.WithAttributes(attrs...))
		}
		if c.completionTokens > 0 {
			attrs := append(append([]attribute.KeyValue{}, genAIAttrs...),
				attribute.String(GenAITokenType, TokenTypeOutput))
			c.rt.clientTokenUsage.Add(ctx, c.completionTokens, metric.WithAttributes(attrs...))
		}

		// Project-specific llm_* dimensions: full tenant bundle so dashboards
		// can slice without joins.
		llmAttrs := llmAttributeSet(c.opts, c.errorKind, c.rt.defaultTags)
		c.rt.requestsTotal.Add(ctx, 1, metric.WithAttributes(llmAttrs...))
		if c.hasUsageDollars {
			c.rt.usageDollars.Add(ctx, c.usageDollars, metric.WithAttributes(llmAttrs...))
		}

		// Reflect token totals + error on the span. We deliberately do not
		// attach prompt/completion text — counts only.
		c.span.SetAttributes(
			attribute.Int64("gen_ai.usage.input_tokens", c.promptTokens),
			attribute.Int64("gen_ai.usage.output_tokens", c.completionTokens),
		)
		if c.errorKind != "" {
			c.span.SetAttributes(attribute.String(AttrErrorType, c.errorKind))
			c.span.SetStatus(codes.Error, c.errorKind)
		} else {
			c.span.SetStatus(codes.Ok, "")
		}
		c.span.End()
	})
}

// buildSpanAttributes assembles the gen_ai.* + multi-tenant attribute set
// recorded on the LLM-call span at start time. Token totals and error class
// are added at End so they reflect final values.
func buildSpanAttributes(opts CallOptions, defaults map[string]string) []attribute.KeyValue {
	attrs := make([]attribute.KeyValue, 0, 12)
	if opts.Provider != "" {
		attrs = append(attrs, attribute.String(GenAISystem, opts.Provider))
	}
	if opts.Model != "" {
		attrs = append(attrs, attribute.String(GenAIRequestModel, opts.Model))
	}
	if opts.Operation != "" {
		attrs = append(attrs, attribute.String(GenAIOperationName, opts.Operation))
	}
	if opts.ServerAddress != "" {
		attrs = append(attrs, attribute.String(AttrServerAddress, opts.ServerAddress))
	}
	if opts.Route != "" {
		attrs = append(attrs, attribute.String(AttrRoute, opts.Route))
	}
	for k, v := range tenantBundle(opts, defaults) {
		attrs = append(attrs, attribute.String(k, v))
	}
	return attrs
}

// genAIAttributeSet returns the GenAI-only attributes recorded on OTel GenAI
// histograms and counters. Multi-tenant fields live on the llm_* counters and
// the span — they do not pollute the gen_ai.* signal.
func genAIAttributeSet(opts CallOptions, errorKind string) []attribute.KeyValue {
	attrs := make([]attribute.KeyValue, 0, 6)
	if opts.Provider != "" {
		attrs = append(attrs, attribute.String(GenAISystem, opts.Provider))
	}
	if opts.Model != "" {
		attrs = append(attrs, attribute.String(GenAIRequestModel, opts.Model))
	}
	if opts.Operation != "" {
		attrs = append(attrs, attribute.String(GenAIOperationName, opts.Operation))
	}
	if opts.ServerAddress != "" {
		attrs = append(attrs, attribute.String(AttrServerAddress, opts.ServerAddress))
	}
	if errorKind != "" {
		attrs = append(attrs, attribute.String(AttrErrorType, errorKind))
	}
	return attrs
}

// llmAttributeSet is the dimension set for llm_requests_total and
// llm_usage_dollars. Multi-tenant fields are first-class.
func llmAttributeSet(opts CallOptions, errorKind string, defaults map[string]string) []attribute.KeyValue {
	attrs := make([]attribute.KeyValue, 0, 10)
	if opts.Provider != "" {
		attrs = append(attrs, attribute.String(AttrProvider, opts.Provider))
	}
	if opts.Model != "" {
		attrs = append(attrs, attribute.String(AttrModel, opts.Model))
	}
	if opts.Route != "" {
		attrs = append(attrs, attribute.String(AttrRoute, opts.Route))
	}
	for k, v := range tenantBundle(opts, defaults) {
		attrs = append(attrs, attribute.String(k, v))
	}
	attrs = append(attrs, attribute.String(AttrErrorKind, errorKind))
	return attrs
}

// tenantBundle merges per-call CallOptions over Options.DefaultTags so the
// per-call value wins. Empty strings never overwrite a default. Returned map
// only contains the multi-tenant keys (tenant/team/app/env/project).
func tenantBundle(opts CallOptions, defaults map[string]string) map[string]string {
	out := make(map[string]string, 5)
	for _, k := range []string{AttrTenant, AttrTeam, AttrApp, AttrEnv, AttrProject} {
		if v, ok := defaults[k]; ok && v != "" {
			out[k] = v
		}
	}
	if opts.Tenant != "" {
		out[AttrTenant] = opts.Tenant
	}
	if opts.Team != "" {
		out[AttrTeam] = opts.Team
	}
	if opts.App != "" {
		out[AttrApp] = opts.App
	}
	if opts.Env != "" {
		out[AttrEnv] = opts.Env
	}
	if opts.Project != "" {
		out[AttrProject] = opts.Project
	}
	return out
}
